### PR TITLE
Improving the SIMD codegen for SIMD12 load/store

### PR DESCRIPTION
**Pull request by tannergooding**
This adds support for containment on TYP_SIMD12 loads/stores and improves the codegen to require less temporary registers and use better instructions when available.

Improved Load:
```diff
- lea      rax, bword ptr [rcx+30H]
- vmovss   xmm0, dword ptr [rax+08H]
- vmovsd   xmm1, qword ptr [rax]
- vshufps  xmm1, xmm0, 68
+ vmovsd   xmm0, qword ptr [rcx+30H]
+ vinsertps xmm0, dword ptr [rcx+38H], 2
```

Improved Store:
```diff
- vmovsd   qword ptr [rdx], xmm1
- vpshufd  xmm0, xmm1, 2
- vmovss   dword ptr [rdx+08H], xmm0
+ vmovsd   qword ptr [rdx], xmm0
+ vextractps dword ptr [rdx+08H], xmm0, 2
```

Combined this saves 9 bytes of codegen and improves the PerScore by 1.5

Total diffs are all relatively similar. Emitting `vmovsd + vinsertps` or `vmovsd + vextractps` and removing now unnecessary `lea` in favor of containing them.

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @dotnet/jit-contrib, this should be ready for review.

Gives some small size savings for x64 (~2k bytes in fullopts and ~0.5k bytes in minopts) and a small TP win on x64

**Comment by tannergooding** [comment]
Fixed the jitstress failure. Results in ~3.3k savings on x86/x64 and a -0.01% TP improvement

File: src/coreclr/jit/emitxarch.cpp [hunk 1]
```diff
@@ -11444,8 +11444,12 @@ BYTE* emitter::emitOutputAM(BYTE* dst, instrDesc* id, code_t code, CnsVal* addc)
             {
                 case IF_AWR_RRD_RRD:
                 {
-                    reg345 = id->idReg2();
-                    break;
+                    if (ins != INS_extractps)
```
Comment (TIHan) [inline src/coreclr/jit/emitxarch.cpp:11447]: This seems like an interesting change. Why is this check here?

**Review by TIHan — COMMENTED** [review]
This looks good to me. There seems to be a failure for ARM though, but it looks like just one test case at the moment.

File: src/coreclr/jit/emitxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/emitxarch.cpp:11447]: There is some existing path that can go down `IF_AWR_RRD_RRD` for `extractps`, even though it should always go down `IF_AWR_RRD_CNS`.

That needs to be addressed separately as part of the larger existing issue covering invalid/nonsensical formats.

**Comment by tannergooding** [comment]
> There seems to be a failure for ARM though, but it looks like just one test case at the moment.

It's an unrelated/existing GC timeout. I've retriggered it and it should pass on rerun.

File: src/coreclr/jit/emitxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/emitxarch.cpp:11447]: - The change involved to properly fix this is a bit more complex/involved and seeing as this is an existing issue, I didn't feel the need to complicate this PR with that.

I plan fixing it in a followup PR alongside a couple other "obvious" places where the format is incorrect.

File: src/coreclr/jit/simdcodegenxarch.cpp [hunk 2]
```diff
@@ -60,40 +60,70 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
...
-    GetEmitter()->emitIns_AR_R(ins_Store(TYP_DOUBLE), EA_8BYTE, data->GetRegNum(), addr->GetRegNum(), 0);
+    if (treeNode->isIndirAddrMode())
+    {
+        GenTreeAddrMode* addrMode = addr->AsAddrMode();
+        addrMode->SetOffset(addrMode->Offset() + 8);
+    }
+    else if (addr->IsCnsIntOrI() && addr->isContained())
+    {
+        GenTreeIntConCommon* icon = addr->AsIntConCommon();
+        icon->SetIconValue(icon->IconValue() + 8);
```
Comment (SingleAccretion) [inline src/coreclr/jit/simdcodegenxarch.cpp:91]: Are we guaranteed to never see relocatable handles here? It is not valid to `icon->IconValue() + 8` for them.

File: src/coreclr/jit/simdcodegenxarch.cpp [hunk 3]
```diff
@@ -104,33 +134,84 @@ void CodeGen::genStoreIndTypeSIMD12(GenTree* treeNode)
...
-    GetEmitter()->emitIns_R_R_I(INS_shufps, emitActualTypeSize(TYP_SIMD16), targetReg, tmpReg, (int8_t)SHUFFLE_YXYX);
+    if (treeNode->isIndirAddrMode())
+    {
+        GenTreeAddrMode* addrMode = addr->AsAddrMode();
+        addrMode->SetOffset(addrMode->Offset() + 8);
+    }
+    else if (addr->IsCnsIntOrI() && addr->isContained())
+    {
+        GenTreeIntConCommon* icon = addr->AsIntConCommon();
+        icon->SetIconValue(icon->IconValue() + 8);
```
Comment (SingleAccretion) [inline src/coreclr/jit/simdcodegenxarch.cpp:164]: Ditto: relocatable handles.

File: src/coreclr/jit/simdcodegenxarch.cpp [hunk 4]
```diff
@@ -104,33 +134,84 @@ void CodeGen::genStoreIndTypeSIMD12(GenTree* treeNode)
...
+
+        if (treeNode->isIndirAddrMode())
+        {
+            GenTreeAddrMode* addrMode = addr->AsAddrMode();
+            addrMode->SetOffset(addrMode->Offset() - 8);
+        }
+        else if (addr->IsCnsIntOrI() && addr->isContained())
+        {
+            GenTreeIntConCommon* icon = addr->AsIntConCommon();
+            icon->SetIconValue(icon->IconValue() - 8);
```
Comment (SingleAccretion) [inline src/coreclr/jit/simdcodegenxarch.cpp:199]: Ditto: relocatable handles.

File: src/coreclr/jit/simdcodegenxarch.cpp (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/simdcodegenxarch.cpp:91]: Discussed this with Single over Discord. Pretty sure we can't hit this today, but might in the future with some planned NAOT optimizations.

It was trivial to workaround, however, and has been handled accordingly.

Comment (tannergooding) [reply src/coreclr/jit/simdcodegenxarch.cpp:92]: I had thought relocatable handles were always blocked from containment, but that isn't true for indirs.

**Review by TIHan — APPROVED** [review]

**Pull request merged**